packages/autocomplete-plugin-tags/src/createTagsPlugin.tsx

```diff
@@ -0,0 +1,173 @@
+  /**
+   * Adds tags to the list.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-addtags
+   */
+  addTags: (tags: Array<BaseTag<TTag>>) => void;
+  /**
+   * Sets the list of tags.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-settags
+   */
+  setTags: (tags: Array<BaseTag<TTag>>) => void;
+};
+
+export type CreateTagsPluginParams<
+  TTag extends Record<string, any>,
+  TItem extends BaseItem
+> = {
+  /**
+   * A set of initial tags to pass to the plugin.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-initialtags
+   */
+  initialTags?: Array<BaseTag<TTag>>;
+  /**
+   * A function to specify what sources the plugin should subscribe to. The plugin adds a tag when selecting an item from these sources.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-gettagssubscribers
+   */
+  getTagsSubscribers?(): Array<{
+    sourceId: string;
+    getTag(params: { item: TItem }): BaseTag<TTag>;
+  }>;
+  /**
+   * A function to transform the returned tags source.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-transformsource
+   */
+  transformSource?(params: {
+    source: AutocompleteSource<Tag<TTag>>;
+    state: AutocompleteState<Tag<TTag>>;
+  }): AutocompleteSource<Tag<TTag>> | undefined;
+  /**
+   * A function called when the list of tags changes.
+   *
+   * @link https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-plugin-tags/createtagsplugin/#param-onchange
+   */
+  onChange?(params: OnChangeParams<TTag>): void;
+};
+
+export function createTagsPlugin<TTag, TItem extends BaseItem>({
+  initialTags = [],
+  getTagsSubscribers = () => [],
+  transformSource = ({ source }) => source,
+  onChange = () => {},
+}: CreateTagsPluginParams<TTag, TItem> = {}): AutocompletePlugin<
+  Tag<TTag>,
+  TagsPluginData<TTag>
+> {
+  const tags = createTags({ initialTags });
+  const tagsApi = { setTags: tags.set, addTags: tags.add };
+
+  return {
+    subscribe(params) {
+      const { setContext, onSelect, setIsOpen, refresh } = params;
+      const subscribers = getTagsSubscribers();
+
+      setContext({ tagsPlugin: { ...tagsApi, tags: tags.get() } });
+
+      onSelect(({ source, item }) => {
+        const subscriber = subscribers.find(
+          ({ sourceId }) => sourceId === source.sourceId
+        );
+
+        if (subscriber) {
+          tags.add([subscriber.getTag({ item })]);
+        }
+      });
+
+      tags.onChange(({ prevTags }) => {
+        setContext({
+          tagsPlugin: { ...tagsApi, tags: tags.get() },
+        });
+
+        setIsOpen(true);
```

Comment:
Opening the panel makes sense when you display tags in a source, but it doesn't when you display them in the search box. Is it a good default?

Reply:
I don't think this is only about where tags are displayed, but about what you do with them. If you use tags for filtering, you might want to see your results update. Do you have a use case in mind where it wouldn't be desirable?

Reply:
I think it is, yes. This isn't about where tags are displayed, but about what you do with them. In the vast majority of cases, tags will likely be used to derive Algolia `filters` (whether it's for actual filtering or for browsing)—and even if they're not, the idea is for tags to have some kind of influence on results. Therefore, you do need to see your reuslts update, it doesn't make sense to close the panel after applying a tag.

If users do want to close the panel, they can use `setIsOpen` in `onChange`.

Reply:
When you're using an external search results page for instance (with InstantSearch), you want the panel to close because adding a tag is the final action.

As long as we can call `setIsOpen(false)` user-land, we're good!

Reply:
Yes we can!

Not sure I get the InstantSearch example. What's the final action?

Reply:
The final action is to add a tag, which would close the panel and apply a filter to the InstantSearch results (and not refresh the Autocomplete results).

Reply:
Ah, got it. Yes, it's possible userland :)